engine/crates/composition/src/compose/enums.rs

```diff
@@ -78,41 +77,58 @@ fn merge_intersection(
         ));
     }
 
-    let enum_id = ctx.insert_enum(first.name(), is_inaccessible, description);
+    let enum_id = ctx.insert_enum(first.name(), description, composed_directives);
 
     for value in intersection {
-        let deprecation = definitions
+        let containers = definitions
             .iter()
-            .find_map(|def| {
-                def.enum_value_by_name(value)
-                    .and_then(|val| val.directives().deprecated())
-            })
-            .map(|dep| dep.reason());
-
-        ctx.insert_enum_value(enum_id, first.walk(value), deprecation, None);
+            .filter_map(|enm| enm.enum_value_by_name(value))
+            .map(|value| value.directives());
+        let composed_directives = collect_composed_directives(containers, ctx);
+        ctx.insert_enum_value(enum_id, first.walk(value), None, composed_directives);
     }
 }
 
-fn merge_union(
-    first: &DefinitionWalker<'_>,
-    definitions: &[DefinitionWalker<'_>],
-    is_inaccessible: bool,
-    ctx: &mut Context<'_>,
+fn merge_union<'a>(
+    first: &DefinitionWalker<'a>,
+    definitions: &[DefinitionWalker<'a>],
+    composed_directives: Vec<federated::Directive>,
+    ctx: &mut Context<'a>,
 ) {
     let description = definitions.iter().find_map(|def| def.description());
-    let enum_id = ctx.insert_enum(first.name(), is_inaccessible, description);
+    let enum_id = ctx.insert_enum(first.name(), description, composed_directives);
+    let mut all_values: Vec<(StringId, _)> = definitions
+        .iter()
+        .flat_map(|def| def.enum_values().map(|value| (value.name().id, value.directives().id)))
+        .collect();
+
+    all_values.sort();
+
+    let mut start = 0;
 
-    for value in definitions.iter().flat_map(|def| def.enum_values()) {
-        let deprecation = value.directives().deprecated().map(|dep| dep.reason());
-        ctx.insert_enum_value(enum_id, value.name(), deprecation, None);
+    loop {
+        let name = all_values[start].0;
+        let end = all_values[start..].partition_point(|(n, _)| *n == name) + start;
+        let containers = all_values[start..end]
+            .iter()
+            .map(|(_, directives)| first.walk(*directives));
+        let composed_directives = collect_composed_directives(containers, ctx);
+
+        ctx.insert_enum_value(enum_id, first.walk(name), None, composed_directives);
+
+        if end == all_values.len() {
```

Comment:
`end` could be mutable and you would be able to do a simpler `while end < all_value.len()` loop

Reply:
or what you did later with mutable `start` :D